pandas/tests/reshape/merge/test_merge.py

```diff
@@ -2094,3 +2094,33 @@ def test_merge_equal_cat_dtypes2():
 
     # Categorical is unordered, so don't check ordering.
     tm.assert_frame_equal(result, expected, check_categorical=False)
+
+
+def test_merge_multiindex_columns():
+    # Issue #28518
+    # Verify that merging two dataframes give the expected labels
+    # The original cause of this issue come from a bug lexsort_depth and is tested in
+    # test_lexsort_depth
+
+    index_tuples = []
+    letters = ["a", "b", "c", "d"]
+    numbers = ["1", "2", "3"]
+
+    for l in letters:
+        for n in numbers:
+            index_tuples.append([l, n])
+
+    index = pd.MultiIndex.from_tuples(index_tuples, names=["outer", "inner"])
```

Comment:
Can you not just use `from_product` here instead to replace the two loops above?

Reply:
Yes cleary.

Reply:
Yes clearly.